### PR TITLE
Nonblocking send/recv

### PR DESCRIPTION
**Pull request by ye-luo**
This PR introduce two optimization in the walker send/recv. This is the third PR for #609.
1, multiple copy of one walker is sent in one shot. This reduces the amount of data transferred.
2, Using non-blocking send/recv to avoid blocking send/recv serialization. Turned on by default.
Both algorithms will be transferred to reconfiguration in the near future.

Command line option async_swap has been removed as well as the old implementation.
input parameter use_nonblocking is introduced to toggle on/off non-blocking send/recv.
manual has been updated.

### CODE REVIEW

File: src/QMCDrivers/DMC/WalkerControlMPI.cpp [hunk 1]
```diff
@@ -190,183 +191,203 @@ void WalkerControlMPI::swapWalkersSimple(MCWalkerConfiguration& W)
...
   fout << std::endl;
 #endif
- //  From the class
-  // MyContext (read)
-  // myComm (read)
-  // NumWalkersSent (write)
-  int nswap=std::min(plus.size(), minus.size());
+  if(plus.size()!=minus.size())
+  {
+    app_error() << "Walker send/recv pattern doesn't match." << std::endl;
```
Comment (markdewing) [inline src/QMCDrivers/DMC/WalkerControlMPI.cpp]: Could you put the plus and minus sizes in the error message.

**Comment by prckent** [comment]
(I expect to review this later today)

**Comment by ye-luo** [comment]
OpenMP is not in send/recv. It is only in walker copy. This is checked by a DMC with no MPI build.
send/recv part is MPI only but MPI probably uses threads to handle non blocking send/recv.

**Comment by prckent** [comment]
Yes, we have several tests that check this. But I felt it was worth noting that a pure MPI run would no longer catch everything.

**Pull request merged**